src/delta/delta.c

```diff
@@ -426,7 +437,12 @@ static int process_suffix(const char *suffix, const char *onlyprefix) {
 
 finish:
         ordered_hashmap_free_free(top);
-        ordered_hashmap_free_free(bottom);
+
+        ORDERED_HASHMAP_FOREACH_KEY(h, key, bottom) {
+                free(ordered_hashmap_remove(bottom, key));
+                free(key);
+        }
```

Comment:
so, yeah, another thing you need to fix, when touching this: please switch this hash map to use hash_ops that free key/value. Then switch away from `ordered_hashmap_free_free()` to regular `ordered_hashmap_free()`.

(yes, patching though the entire codebase also means a lot of extra work to update relevant code pieces to newest coding style, sorry!)

(in case this wasn't clear: we want to kill ordered_hashmap_free_free() and manual destrction loops, and instead want that destructors are defined in the hash_ops structure, everywhere. Much of the codebase has been converted, but there are some hold-outs. Just like with path_extract_filename() itself)

Reply:
Should that be in a separate commit or all at once?

Reply:
i don't mind either way. we prefer smaller commits up to a certain point. i think a separate commit would be fine here, but doing it in one might be fine too

Reply:
Something I can't immediately grok from the source: does `ordered_hashmap_replace` and/or `ordered_hashmap_remove` free the key/value or would that still need to manually be done?  The only place I could see the `free_key` being called grepping `src/basic/hashmap.c` is in `ordered_hashmap_clear` and `ordered_hashmap_free`, but I could be missing it.

Reply:
The hash_ops structure contains key and value destructors.

Reply:
I have this mostly working, but it's still leaking a bit, and trying to debug that has led me to a _very_ interesting valgrind bug.  So it might be a few days before I can finish this up…

Reply:
I rewrote this mostly correct.  However, valgrind showed:

```
==12440== HEAP SUMMARY:
==12440==     in use at exit: 4,154 bytes in 3 blocks
==12440==   total heap usage: 1,108 allocs, 1,105 frees, 64,473 bytes allocated
==12440== 
==12440== 18 bytes in 1 blocks are definitely lost in loss record 1 of 3
==12440==    at 0x4919448: malloc (in /usr/libexec/valgrind/vgpreload_memcheck-ppc64be-linux.so)
==12440==    by 0x4086EBB: strdup (strdup.c:7)
==12440==    by 0x10CC9F: enumerate_dir (delta.c:362)
==12440==    by 0x10D513: process_suffix (delta.c:452)
==12440==    by 0x10D83F: process_suffixes (delta.c:494)
==12440==    by 0x10E41B: run (delta.c:674)
==12440==    by 0x10E573: main (delta.c:686)
```

This is because the `free(ordered_hashmap_remove(bottom, bp));` doesn't free the key.  I needed to use `ordered_hashmap_remove2` (cribbing from sd-device.c).  Leaving this note here in case anyone else is doing this modernisation work and can't figure it out.